1.regression_modelling/scripts/1.train_models.py

```diff
@@ -0,0 +1,213 @@
+                if len(parts) > 4 and ("A647" in parts[3] or "A647" in parts[4]):
+                    a647_features.append(column)
+                elif len(parts) > 4 and ("GOLD" in parts[3] or "GOLD" in parts[4]):
+                    gold_features.append(column)
+            else:  # No DAPI in correlation feature, check only 4th part
+                if len(parts) > 3 and "A647" in parts[3]:
+                    a647_features.append(column)
+                elif len(parts) > 3 and "GOLD" in parts[3]:
+                    gold_features.append(column)
+        else:  # Non-correlation features
+            if len(parts) > 4 and "Location" in parts[1]:  # If it's a Location feature
+                if parts[4] == "DAPI":
+                    nucleus_features.append(column)
+                elif parts[4] == "A647":
+                    a647_features.append(column)
+                elif parts[4] == "GOLD":
+                    gold_features.append(column)
+            elif len(parts) > 3 and "DAPI" in parts[3]:
+                nucleus_features.append(column)
+            elif len(parts) > 3 and "A647" in parts[3]:
+                a647_features.append(column)
+            elif len(parts) > 3 and "GOLD" in parts[3]:
+                gold_features.append(column)
+            else:
+                nucleus_features.append(column)  # Default to nucleus_features
+
+
+# Prepare X data for with all nucleus features
+X = training_df[nucleus_features]
+
+# Generate shuffled data for the shuffled models to use (only do this once)
+X_shuffled = X.copy()
+for col in X_shuffled.columns:
+    np.random.shuffle(X_shuffled[col].values)  # Shuffle values in place, independently
+
+# Print the lists to verify
+print(f"Nucleus Features: {len(nucleus_features)}")
+print(f"A647 Features: {len(a647_features)}")
+print(f"Gold Features: {len(gold_features)}")
+
+
+# ## Set hyperparameter parameters and search space
+
+# In[4]:
+
+
+# Set folds for k-fold cross-validation
+k_folds = KFold(n_splits=5, shuffle=True, random_state=0)
+
+# Set ElasticNet regression model parameters
+elasticnet_params = {
+    "alpha": 1.0,  # Equivalent to 'C' in LogisticRegression, but in reverse
+    "l1_ratio": 0.5,  # Mixture of L1 and L2 regularization
+    "max_iter": 10,
+    "random_state": 0,
+}
+
+# Define the hyperparameter search space for RandomizedSearchCV
+param_dist = {
+    "alpha": np.logspace(-3, 3, 7),  # Regularization strength
+    "l1_ratio": np.linspace(0, 1, 11),  # Mix of L1 and L2 regularization
+}
+
+# Set the random search hyperparameterization method parameters
+random_search_params = {
+    "param_distributions": param_dist,
+    "scoring": "neg_mean_squared_error",  # Suitable for regression
+    "random_state": 0,
+    "n_jobs": -1,
+    "cv": k_folds,
+}
+
+
+# ## Train A647 models
+
+# In[5]:
+
+
+# Suppress all warnings
+warnings.filterwarnings("ignore")
+
+for a647_feature in a647_features:
+    y = training_df[a647_feature]
+
```

Comment:
I love this approach, we did this for the Pyroptosis paper. I also think in a future PR it might be worth exploring a multi-regression model where Nuclei features predict all the A647 features in one model. You would probably take a hit on performance in fovour of generalizability. This will be up to you if it is worth a future PR. Might be worth it and fit well in a supplementary figure! XD

Reply:
I so agree! It would be really cool to explore so I think this will be kept in mind for a future paper, thank you lots!